Add workaround for observed strange behavior

Michael-Sinz opened the pull request:
Add workaround for observed strange behavior
    
We had a test run that observed an extension still being configured
even after it was deleted successfully from the configuration.
    
This should fix #83 that was reported

To hopefully address this, we re-list the extensions after trying
to delete them and will re-delete those that we feel we should
delete if they still are there after having deleted them.  We
will do that up to 5 times.

The problem is that at this point the new image has been installed
and set up as the target image so the extensions should not be needed
but we can't undo what we did to install new image.  We can only go
forward.  (There is no single operation that can do it all atomically,
especially since now we see that even a successful delete is not
always successful anyway)

Michael-Sinz committed:
Add workaround for observed strange behavior

We had a test run that observed an extension still being configured
even after it was deleted successfully from the configuration.

This was reported in issue/bug #83

To hopefully address this, we re-list the extensions after trying
to delete them and will re-delete those that we feel we should
delete if they still are there after having deleted them.  We
will do that up to 5 times.

The problem is that at this point the new image has been installed
and set up as the target image so the extensions should not be needed
but we can't undo what we did to install new image.  We can only go
forward.  (There is no single operation that can do it all atomically,
especially since now we see that even a successful delete is not
always successful anyway)

jackfrancis commented on vmss-prototype/vmss-prototype:
```diff
@@ -1014,32 +1014,41 @@ def vmss_prototype_update(sub_args):
+                    '--resource-group', resource_group,
+                    '--vmss-name', vmss,
+                    '--name', extension['name']
+                    ], retries=3, check=False, retry_func=not_found_no_retry, log_retry_errors=True)
```
Should we add a sleep at the end of this while loop? In the event that our follow-up `vmss extension list` operation still returns an extension (not the `vmss-computeAksLinuxBilling` one), our 3 remaining retries will happen in quick succession.

Maybe we should add a simple thing at the top of this loop:

```python
if delete_tries < 4:
    sleep 30 # caveat emptor I don't know the correct python sleep statement off the top of my head :)
```

Michael-Sinz replied:
Not a bad idea - to sleep based on how much we have done.  Let me look into it.

Michael-Sinz replied:
However, the tricky bit is that I really don't want to sleep if we have correctly deleted everything.  We always loop twice, the second time just notices that all is deleted.

Michael-Sinz replied:
Note that the backoff is likely redundant since it takes seconds (like 10+ seconds) when an extension is deleted (as per the logs I captured) so it was already somewhat slow roll already.

However, it can not hurt to put the backoff logic in to help.

Michael-Sinz committed:
Add backoff timing to retry loop

Also minor cleanups and updates to my smoketests to run on control
plane such that they work in generic aks-engine clusters

Michael-Sinz commented:
Note - looking at the check/test logs, it does not seem to be running the new code (the code that is part of the PR)

It clearly did build something but I could not follow what it built - I think it build the wrong container image.

My local testing showed that the loop does do the second list of extensions to validate that any that would have been removed are now removed.  I even ran another test here to validate.

```
2021-09-23T14:34:44.535647205Z k8s-agentpool1-00000002-vmss INFO: ===> Executing command: ['az' 'vmss' 'update' '--subscription' '00000000-0000-0000-0000-000000000001' '--resource-group' 'testing' '--name' 'k8s-agentpool1-00000002-vmss' '--set' 'virtualMachineProfile.storageProfile.imageReference.id=/subscriptions/00000000-0000-0000-0000-000000000001/resourceGroups/testing/providers/Microsoft.Compute/galleries/SIG_testing/images/kamino-k8s-agentpool1-00000002-vmss-prototype' 'virtualMachineProfile.storageProfile.imageReference.sku=null' 'virtualMachineProfile.storageProfile.imageReference.offer=null' 'virtualMachineProfile.storageProfile.imageReference.publisher=null' 'virtualMachineProfile.storageProfile.imageReference.version=null' 'virtualMachineProfile.osProfile.customData=I2Nsb3VkLWNvbmZpZwo=']
2021-09-23T14:34:57.957465948Z k8s-agentpool1-00000002-vmss INFO: ===> Executing command: ['az' 'vmss' 'show' '--subscription' '00000000-0000-0000-0000-000000000001' '--resource-group' 'testing' '--name' 'k8s-agentpool1-00000002-vmss']
2021-09-23T14:34:59.243837134Z k8s-agentpool1-00000002-vmss INFO: ===> Executing command: ['az' 'vmss' 'extension' 'list' '--subscription' '00000000-0000-0000-0000-000000000001' '--resource-group' 'testing' '--vmss-name' 'k8s-agentpool1-00000002-vmss']
2021-09-23T14:35:00.488459855Z k8s-agentpool1-00000002-vmss INFO: ===> Executing command: ['az' 'vmss' 'extension' 'delete' '--subscription' '00000000-0000-0000-0000-000000000001' '--resource-group' 'testing' '--vmss-name' 'k8s-agentpool1-00000002-vmss' '--name' 'vmssCSE']
2021-09-23T14:35:13.986576542Z k8s-agentpool1-00000002-vmss INFO: ===> Executing command: ['az' 'vmss' 'extension' 'list' '--subscription' '00000000-0000-0000-0000-000000000001' '--resource-group' 'testing' '--vmss-name' 'k8s-agentpool1-00000002-vmss']
```
